nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/CopyAzureBlobStorage_v12.java

```diff
@@ -0,0 +1,330 @@
+                      final BlobRequestConditions sourceRequestConditions,
+                      final BlobRequestConditions destinationRequestConditions) {
+        BlockBlobClient blockBlobClient = blobClient.getBlockBlobClient();
+
+        // If the blob size is below the limit, we use the one-shot upload endpoint.
+        if (blobSize < MAX_UPLOAD_BLOB_BYTES_LONG) {
+            final BlobUploadFromUrlOptions options = new BlobUploadFromUrlOptions(sourceUrl);
+            if (httpAuthorization != null) {
+                options.setSourceAuthorization(httpAuthorization);
+            }
+            options.setSourceRequestConditions(sourceRequestConditions);
+            options.setDestinationRequestConditions(destinationRequestConditions);
+            blockBlobClient.uploadFromUrlWithResponse(options, null, Context.NONE);
+            return;
+        }
+
+        final DecimalFormat df = new DecimalFormat("0000000");
+        long offset = 0;
+        int blockId = 1;
+        final List<String> blockIds = new ArrayList<>();
+
+        // Upload each block sequentially until we've uploaded the entire blob.
+        while (true) {
+            long count = Math.min(blobSize - offset, MAX_STAGE_BLOCK_BYTES_LONG);
+            if (count == 0) break;
+
+            final String base64BlockId = Base64.getEncoder().encodeToString(df.format(blockId).getBytes());
+            BlockBlobStageBlockFromUrlOptions blockBlobStageBlockFromUrlOptions = new BlockBlobStageBlockFromUrlOptions(base64BlockId, sourceUrl);
+            blockBlobStageBlockFromUrlOptions.setSourceRange(new BlobRange(offset, count));
+
+            if (httpAuthorization != null) {
+                blockBlobStageBlockFromUrlOptions.setSourceAuthorization(httpAuthorization);
+            }
+            blockBlobStageBlockFromUrlOptions.setSourceRequestConditions(sourceRequestConditions);
+            final int statusCode = blockBlobClient.stageBlockFromUrlWithResponse(blockBlobStageBlockFromUrlOptions, null, Context.NONE).getStatusCode();
+            if (statusCode != 201) {
+                throw new ProcessException(String.format("Failed staging one or more blocks (status: %d)", statusCode));
+            }
+            blockIds.add(base64BlockId);
+            offset += count;
+            blockId++;
+        }
+
+        final BlockBlobCommitBlockListOptions options = new BlockBlobCommitBlockListOptions(blockIds);
+        options.setRequestConditions(destinationRequestConditions);
+        final Response<BlockBlobItem> response = blockBlobClient.commitBlockListWithResponse(options, null, Context.NONE);
+        final int statusCode = response.getStatusCode();
+        if (statusCode != 201) {
+            throw new ProcessException(String.format("Failed committing block list (status: %d)", statusCode));
+        }
+    }
+
+    private static HttpAuthorization getHttpAuthorization(final AzureStorageCredentialsDetails_v12 credentialsDetails) {
```

Comment:
I feel like this authentication-related code shouldn't belong here because we have something similar in the abstract class. After a quick glance, I don't know a good solution, but it is worth investigating to merge those code parts somehow.

Reply:
I have tried but there are some major differences and it didn't seem like it was really possible to merge them last time I checked.

Reply:
@nandorsoma in BlobServiceClientFactory, we have a method `configureCredential` which configures a `BlobServiceClientBuilder`. But when copying, the source is not a client really, we're just adding some authorization headers. That is, there's a bit of a mismatch between the two methods even though there's some shared logic.

I don't think it can be refactored easily.